eng/packages/http-client-csharp/emitter/src/emitter.ts

```diff
@@ -1,18 +1,98 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
-import { EmitContext } from "@typespec/compiler";
+import { EmitContext, Program, resolvePath, } from "@typespec/compiler";
+import { createSdkContext, DecoratorInfo } from "@azure-tools/typespec-client-generator-core";
 
 import {
-  $onEmit as $OnMGCEmit,
   CSharpEmitterOptions,
-  setSDKContextOptions
+  createModel,
+  InputModelType,
+  Logger,
+  LoggerLevel,
+  resolveOptions,
+  emit
 } from "@typespec/http-client-csharp";
 import { azureSDKContextOptions } from "./sdk-context-options.js";
+import {
+  armResourceCreateOrUpdate,
+  armResourceOperations,
+  armResourceRead,
+  singleton,
+  resourceMetadata,
+} from "./knownDecorators.js";
+import { CalculateResourceTypeFromPath } from "./resource-type.js";
 
 export async function $onEmit(context: EmitContext<CSharpEmitterOptions>) {
+  const program = context.program
   context.options["plugin-name"] ??= "AzureClientPlugin";
   context.options["emitter-extension-path"] = import.meta.url;
-  setSDKContextOptions(azureSDKContextOptions);
-  await $OnMGCEmit(context);
+
+  
+  const options = resolveOptions(context);
+  /* set the log level. */
+  const logger = new Logger(program, options.logLevel ?? LoggerLevel.INFO);
+  
+  // Write out the dotnet model to the output path
+  const sdkContext = {
+    ...(await createSdkContext(
+      context,
+      "@typespec/http-client-csharp",
+      azureSDKContextOptions,
+    )),
+    logger: logger,
+    __typeCache: {
+      crossLanguageDefinitionIds: new Map(),
+      types: new Map(),
+      models: new Map(),
+      enums: new Map(),
+    },
+  };
+  program.reportDiagnostics(sdkContext.diagnostics);
+
+  const root = createModel(sdkContext);
+  if (!root) {
+    return;
+  }
+
+  for (let client of root.Clients) {
+    // TODO: we can implement this decorator in TCGC until we meet the corner case
+    // if the client has resourceMetadata decorator, it is a resource client and we don't need to add it again
+    if (client.Decorators?.some( d => d.name == resourceMetadata)) {
+      continue;
+    }
+    
+    // TODO: Once we have the ability to get resource hierarchy from TCGC directly, we can remove this implementation
+    // A resource client should have decorator armResourceOperations and contains either a get operation(containing armResourceRead deocrator) or a put operation(containing armResourceCreateOrUpdate decorator)
+    if (client.Decorators?.some(d => d.name == armResourceOperations) && client.Operations.some(op => op.Decorators?.some(d => d.name == armResourceRead || armResourceCreateOrUpdate))) {
```

Comment:
Why are we implementing this in the emitter rather than the generator?

Reply:
Eventually, we would like to get resource hierarchy from TypeSpec natively via an API in https://github.com/Azure/azure-sdk-for-net/issues/48281.
After discussion with @m-nash, we would like to keep the contract for generator unchanged. So that, we decide to implement this in emitter to keep the contract stable for generator.
And when we call the tsp API to get resource hierarchy, we will need to do it in emitter here.

Reply:
Sure.
The input type library is not extensible for sub-plugins. The only way to extend it is via decorators.
Let's say we have an API in place from tsp package that we can directly get resource hierarchy. We will still need to pass the resource hierarchy metadata via decorators.
That's why we are preparing the resource metadata in emitter, and put decorators in to the client model.
So that, generator will consume the information directly, just like we have an API in place.
Once we have the API ready, we can do a simple switch in the eimtter.